### ISSUE TITLE
Akka MIME unmarshallers neither accept nor properly reject MIME with LF line endings

### ISSUE THREAD

**Issue by RichardBradley** [akka/akka#17110]
The Akka MIME unmarshallers do not accept (nor properly reject) MIME with LF line endings.

I know that unix line endings are not valid MIME according to the RFC, but Spray 1.3 used to accept this kind of MIME (via the `org.jvnet.mimepull` library), and Postel's law would seem to apply.

Currently, if you pass Akka a MIME message with LF line endings, it parses successfully as a MIME bundle with zero BodyParts. I think that it should either be an error as the RFC states "[The body must then contain **one or more** body parts](http://www.w3.org/Protocols/rfc1341/7_2_Multipart.html)", or (much better) it should parse the slightly broken MIME.

Here’s a failing test demonstrating the issue. The equivalent test in Spray 1.3.1 passes:

```
import akka.actor.ActorSystem
import akka.http.marshalling.Marshal
import akka.http.model.ContentTypes._
import akka.http.model.{HttpEntity, Multipart}
import akka.http.model.Multipart.FormData
import akka.http.unmarshalling.Unmarshal
import akka.stream.ActorFlowMaterializer
import akka.stream.scaladsl.Source
import akka.util.ByteString
import akka.http.marshalling.MultipartMarshallers._
import org.specs2.mutable.Specification
import org.specs2.time.NoTimeConversions
import scala.concurrent.{Future, Await}
import scala.concurrent.duration._

class MimeLineEndingsSpec
  extends Specification
  with NoTimeConversions {

  implicit val system = ActorSystem()
  implicit val fm = ActorFlowMaterializer()
  val timeout = 1.second
  private def await[T](future: Future[T]): T =
    Await.result(future, timeout)

  "MIME unmarshallers" should {
    "accept LF endings as well as CRLF" in {
      val unmarshalledMime = Multipart.FormData(Source(List(
        FormData.BodyPart("first",
          HttpEntity.Strict(`text/plain`, ByteString("first part"))),
        FormData.BodyPart("second",
          HttpEntity.Strict(`text/plain`, ByteString("second part"))))))

      val marshalledMime = await(
        Marshal(unmarshalledMime).to[HttpEntity].flatMap(_.toStrict(timeout)))

      println(marshalledMime.data.utf8String)

      val mimeWithUnixLineEndings = marshalledMime.copy(
        data = ByteString(marshalledMime.data.utf8String.replace("\r\n", "\n")))

      val unmarshalledMimeWithUnixLineEndings = await(
        Unmarshal(mimeWithUnixLineEndings).to[Multipart.FormData])

      // unmarshalledMimeWithUnixLineEndings here is 'FormData.Strict(Vector())'
      // i.e. the whole body is treated as a MIME preamble and ignored.
      //
      // Shouldn't we get an error like "no MIME boundary found"?
      // Or better (and like Spray 1.3), shouldn't we be able to parse LF MIME?
      unmarshalledMimeWithUnixLineEndings mustEqual unmarshalledMime
    }
  }
}
```

Thanks,

Rich

**Comment by RichardBradley** [comment]
// cc @sirthias @jrudolph

**Comment by sirthias** [comment]
Thanks, Richard, for this report.
You identified two issues:

1. The spec disallows zero body-parts while we currently accept this.
2. We should be a bit more lenient with regard to line endings.

I agree with both point and will put in a PR shortly.

**Comment by RichardBradley** [comment]
Thanks very much.

(If you're changing the MIME parser - you might like to see if you can shed any light on #17051 -- I couldn't see where that issue arose in the code when I looked.)

**Comment by sirthias** [comment]
Actually, the second point is somewhat tricky.
We currently use a simple Boyer-Moore string search implementation for efficiently detecting the boundaries in the MIME content. Unfortunately, Boyer-Moore is an *exact* string search algorithm. It cannot easily look for several patterns at once.
We *could* of course always run the boundary search with two patterns at the same time (once for CRLF line endings and once for just LF endings). However, this would decrease parsing performance by more than 50%, which is not something that I'd like to accept just to accommodate for actually illegal content.
Unless someone has an idea how we can support both types of line endings at the same time without significantly increasing implementation complexity or runtime overhead I'd vote for only fixing the first issue.

**Comment by RichardBradley** [comment]
Good point. Let me check how `org.jvnet.mimepull` handled this

**Comment by sirthias** [comment]
Thanks for the heads up to #17051, @RichardBradley!
I'll get that fixed next.

**Comment by sirthias** [comment]
Arrg. Actually the pattern search is easy to adapt to both line endings because the line ending appears at the very beginning of the search pattern. So the ability to search for both line endings at the same time is actually really easy to do.
Sorry for the noise!

**Comment by sirthias** [comment]
Ok, just attempted line ending leniency in MIME parsing.
This is not hard to achieve in the BodyPartParser itself but hits another issue, which is more difficult to solve: The `BodyPartParser` delegates header parsing to the `HttpHeaderParser` which also performs header parsing for ordinary requests and responses. This header parser is quite optimized and has the CRLF treatment hard-coded.
Theoretically it'd be possible to change this is but for my taste it's not worth the effort just for supporting off-spec input.
So, I'd like to (again, but this time for real) reject the change to support non-CRLF line endings.

**Comment by RichardBradley** [comment]
That seems sensible. The header parsing in Spray 1.3 was done by `org.jvnet.mimepull` via a different code path to "normal" HTTP header parsing. (I don't know whether Spray allowed \n delimited headers outside of MIME.)

However, I think it would be an improvement to support LF endings here and also in regular headers.

Have a look at [RFC 2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec19.html#sec19.3):

> The line terminator for message-header fields is the sequence CRLF. However, we recommend that applications, when parsing such headers, recognize a single LF as a line terminator and ignore the leading CR.

From a quick look at `HttpHeaderParser`, can we not fix this just by changing `scanHeaderValue` at line 488 to look for `\r?\n` as the terminator?

**Comment by sirthias** [comment]
> That seems sensible. The header parsing in Spray 1.3 was done by org.jvnet.mimepull via a different code path to "normal" HTTP header parsing. (I don't know whether Spray allowed \n delimited headers outside of MIME.)

No, it doesn't.

> However, I think it would be an improvement to support LF endings here and also in regular headers.
> Have a look at RFC 2616: ...

Yes, agreed. (The relevant spec is now [RFC 7230 Section 3.5](http://tools.ietf.org/html/rfc7230#section-3.5) though.)
Still, we have to prioritize any work in the context of everything that is still on our plate. And, unfortunately, accommodating for irregular peers is not generally a top priority.

> From a quick look at HttpHeaderParser, can we not fix this just by changing scanHeaderValue at line 488 to look for \r?\n as the terminator?

Nope. The line endings are stored in the cache together with the other value bytes.
I'm not saying that the fix comprises a lot of code. It's simply not something that I'd like to approach right now.

Therefore I've just created a ticket for this: https://github.com/akka/akka/issues/17114

**Comment by RichardBradley** [comment]
Thanks, that sounds fine to me

**Comment by sirthias** [comment]
Note that  #17112 is already done to fix the first issue.
The fix to #17051 is in the works.
